driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -0,0 +1,371 @@
+     *
+     * @return true if a timeout has been set.
+     */
+    public boolean hasTimeoutMS() {
+        return timeoutSettings.getTimeoutMS() != null;
+    }
+
+    /**
+     * Checks the expiry of the timeout.
+     *
+     * @return true if the timeout has been set and it has expired
+     */
+    public boolean hasExpired() {
+        // Use timeout.remaining instead of timeout.hasExpired that measures in nanoseconds.
+        return timeout != null && !timeout.isInfinite() && timeout.remaining(MILLISECONDS) <= 0;
+    }
+
+    /**
+     * Sets the recent min round trip time
+     * @param minRoundTripTimeMS the min round trip time
+     * @return this
+     */
+    public TimeoutContext minRoundTripTimeMS(final long minRoundTripTimeMS) {
+        isTrue("'minRoundTripTimeMS' must be a positive number", minRoundTripTimeMS >= 0);
+        this.minRoundTripTimeMS = minRoundTripTimeMS;
+        return this;
+    }
+
+    public Optional<MongoOperationTimeoutException> validateHasTimedOutForCommandExecution() {
+        if (hasTimedOutForCommandExecution()) {
+            return Optional.of(createMongoTimeoutException());
+        }
+        return Optional.empty();
+    }
+
+    private boolean hasTimedOutForCommandExecution() {
+        if (timeout == null || timeout.isInfinite()) {
+            return false;
+        }
+        long remaining = timeout.remaining(MILLISECONDS);
+        return remaining <= 0 || minRoundTripTimeMS > remaining;
+    }
+
+    /**
+     * Returns the remaining {@code timeoutMS} if set or the {@code alternativeTimeoutMS}.
+     *
+     * @param alternativeTimeoutMS the alternative timeout.
+     * @return timeout to use.
+     */
+    public long timeoutOrAlternative(final long alternativeTimeoutMS) {
+        Long timeoutMS = timeoutSettings.getTimeoutMS();
+        if (timeoutMS == null) {
+            return alternativeTimeoutMS;
+        }
+        return timeoutRemainingMS();
+    }
+
+    /**
+     * Calculates the minimum timeout value between two possible timeouts.
+     *
+     * @param alternativeTimeoutMS the alternative timeout
+     * @return the minimum value to use.
+     */
+    public long calculateMin(final long alternativeTimeoutMS) {
+        Long timeoutMS = timeoutSettings.getTimeoutMS();
+        if (timeoutMS == null) {
+            return alternativeTimeoutMS;
+        } else if (timeoutMS == 0) {
+            return alternativeTimeoutMS;
+        } else if (alternativeTimeoutMS == 0) {
+            return timeoutRemainingMS();
+        } else {
+            return Math.min(timeoutRemainingMS(), alternativeTimeoutMS);
+        }
+    }
+
+    public TimeoutSettings getTimeoutSettings() {
+        return timeoutSettings;
+    }
+
+    public long getMaxAwaitTimeMS() {
+        return timeoutSettings.getMaxAwaitTimeMS();
+    }
+
+    public long getMaxTimeMS() {
+        return notNull("Should never be null", maxTimeSupplier.get());
```

Comment:
This method is used only to set the `maxTimeMS` value in `CommandMessage#getExtraElements`, when non-zero. The value is:

1. By default, `calculateMaxTimeMS`: if the timeout not present, use the legacy `maxTimeMS` setting. Otherwise, shorten the timeout by the round trip time, throwing a timeout exception as needed.
2. In `CommitTransactionOperation`, the supplier is overridden to use `getMaxCommitTimeMS`, which uses the timeout if present, but otherwise uses the legacy `maxCommitTimeMS` setting.
3. Elsewhere, the supplier is overridden to return a fixed value of 0, denoting infinite, or to a fixed value like `MaxAwaitTimeMS`.

In the second path, when the `timeout` is present, it is not shortened by the round trip time. In the first, it is shortened, which seems correct. Should it be shortened in the second path?

In the third path, the value (0, or maxAwaitTime) is not overriden by timeout. In the first and second paths, fixed values are overridden by the timeout. Should they be overridden on this third path?

Reply:
Thank you for the detailed question regarding the maxTimeMS settings across different contexts.

**First and Second Paths:** According to the "Command Execution" section in the [MongoDB client-side operations timeout specification](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#command-execution), `maxTimeMS` should be adjusted by the round trip time (RTT). This adjustment is correctly applied in the first path and should indeed be applied in the second path as well. I will create a ticket to address this oversight.

**Third Path**: The handling of `maxTimeMS` using `maxAwaitTimeMS` specifically for ChangeStreams and TailableAwait cursors is outlined as an exception in the specification:

>  If the maxAwaitTimeMS option is set, drivers MUST use it as the maxTimeMS field on getMore commands."

However, assigning `maxAwaitTimeMS` to `maxTimeMS `as a fixed value without considering RTT and the remaining timeout could lead to the connection churn problem as documented in [JIRA DRIVERS-2868](https://jira.mongodb.org/browse/DRIVERS-2868). This issue indicates a need to adjust `maxTimeMS` by RTT and the remaining timeout in these scenarios as well. We are awaiting a spec change to proceed with this adjustment.

In summary, yes, the `maxTimeMS` value should generally be shortened by the RTT to reflect accurate timeout behavior across different scenarios.